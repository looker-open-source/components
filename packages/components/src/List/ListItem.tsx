/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { CompatibleHTMLProps } from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import styled from 'styled-components'
import React, { FC, ReactNode, useContext, useRef, useState } from 'react'
import { ListItemDetail } from '../List/ListItemDetail'
import { Paragraph } from '../Text'
import { Icon, IconPlaceholder } from '../Icon'
import { HoverDisclosureContext, HoverDisclosure, useHovered } from '../utils'
import { ListItemContext } from './ListItemContext'
import { ListItemLayout } from './ListItemLayout'
import { ListItemLayoutAccessory } from './ListItemLayoutAccessory'
import { ListItemWrapper } from './ListItemWrapper'
import { DensityRamp, Detail, ListItemStatefulProps } from './types'
import { createSafeRel, getDetailOptions, listItemDimensions } from './utils'

export interface ListItemProps
  extends CompatibleHTMLProps<HTMLElement>,
    ListItemStatefulProps {
  /**
   * Determines color of child if child is a string
   */
  color?: string
  /**
   * Determines the sizing and spacing of the item
   * Notes:
   * - This prop is intended for internal components usage (density should be set on a parent List component for external use cases).
   * - If you choose to use this prop on a ListItem directly, it must be consistent across all items for windowing purposes.
   * @private
   */
  density?: DensityRamp
  /*
   * optional extra description
   */
  description?: ReactNode
  /**
   * Detail element placed right of the item children. Prop value can take one of two forms:
   * 1. ReactNode
   * 2. Object with content and options properties
   */
  detail?: Detail
  /**
   * Optional icon placed left of the item children
   */
  icon?: IconNames
  /**
   * Display an icon/logo that is not available on our components list. Use artwork prop with an svg instead of Icon name.
   */
  iconArtwork?: ReactNode
  /**
   * Sets the correct accessible role for the ListItem:
   * - Use **'link'** for items that navigation to another page
   * - Use **'button'** for items that trigger in page interactions, like displaying a dialog
   */
  itemRole?: 'link' | 'button'
  /**
   * If true, text children and description will be truncated if text overflows
   */
  truncate?: boolean
}

interface ListItemLabelProps extends CompatibleHTMLProps<HTMLElement> {
  disabled?: boolean
  itemRole?: 'link' | 'button'
}

export const ListItemLabel = styled.div
  .withConfig<ListItemLabelProps>({
    shouldForwardProp: (prop) => prop !== 'itemRole',
  })
  .attrs<ListItemLabelProps>(({ disabled, itemRole }) => ({
    as: !disabled && itemRole === 'link' ? 'a' : 'button',
  }))<ListItemLabelProps>``

const ListItemInternal: FC<ListItemProps> = (props) => {
  const {
    children,
    className,
    color,
    current,
    density: propsDensity,
    description,
    detail,
    disabled,
    href,
    icon,
    iconArtwork,
    itemRole,
    keyColor,
    onBlur,
    onClick,
    onKeyDown,
    onKeyUp,
    rel,
    selected,
    target,
    truncate,
    ...restProps
  } = props

  const { iconGutter, density: contextDensity } = useContext(ListItemContext)

  const itemDimensions = listItemDimensions(propsDensity || contextDensity)

  const [focusVisible, setFocusVisible] = useState(false)

  const itemRef = useRef<HTMLLIElement>(null)
  const [hovered] = useHovered(itemRef)

  const labelColor = disabled ? 'text1' : color
  const descriptionColor = disabled ? 'text1' : 'text2'
  const iconColor = disabled ? 'text1' : color || 'text1'

  const handleOnBlur = (event: React.FocusEvent<HTMLElement>) => {
    setFocusVisible(false)
    onBlur && onBlur(event)
  }

  const handleOnClick = (event: React.MouseEvent<HTMLElement>) => {
    setFocusVisible(false)
    onClick && onClick(event)
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLElement>) => {
    onKeyDown && onKeyDown(event)
    setFocusVisible(true)
  }

  const handleOnKeyUp = (event: React.KeyboardEvent<HTMLElement>) => {
    onKeyUp && onKeyUp(event)
    setFocusVisible(true)
  }

  const renderedIcon =
    icon || iconArtwork ? (
      <Icon
        artwork={iconArtwork}
        color={iconColor}
        name={icon}
        size={itemDimensions.iconSize}
        mr={itemDimensions.iconGap}
      />
    ) : (
      iconGutter && (
        <IconPlaceholder
          size={itemDimensions.iconSize}
          mr={itemDimensions.iconGap}
        />
      )
    )

  if (disabled && itemRole === 'link') {
    // eslint-disable-next-line no-console
    console.warn(
      'itemRole="link" and disabled cannot be combined - use itemRole="button" if you need to offer a disabled ListItem'
    )
  }

  const renderedChildren =
    typeof children === 'string' ? (
      <Paragraph
        color={labelColor}
        fontSize={itemDimensions.labelFontSize}
        lineHeight={itemDimensions.labelLineHeight}
        truncate={truncate}
      >
        {children}
      </Paragraph>
    ) : (
      children
    )

  const renderedDescription =
    typeof description === 'string' ? (
      <Paragraph
        color={descriptionColor}
        fontSize={itemDimensions.descriptionFontSize}
        lineHeight={itemDimensions.descriptionLineHeight}
        truncate={truncate}
      >
        {description}
      </Paragraph>
    ) : (
      description
    )

  const { accessory, content, hoverDisclosure } = getDetailOptions(detail)

  const renderedDetail = detail && (
    <HoverDisclosure visible={!hoverDisclosure}>
      <ListItemDetail pr={accessory ? itemDimensions.px : '0'}>
        {content}
      </ListItemDetail>
    </HoverDisclosure>
  )

  const LabelCreator: FC<{
    children: ReactNode
    className: string
  }> = ({ children, className }) => (
    <ListItemLabel
      itemRole={itemRole}
      aria-current={current}
      aria-selected={selected}
      className={className}
      disabled={disabled}
      href={href}
      onBlur={handleOnBlur}
      onClick={disabled ? undefined : handleOnClick}
      onKeyDown={handleOnKeyDown}
      onKeyUp={handleOnKeyUp}
      rel={createSafeRel(rel, target)}
      role="listitem"
      target={target}
      tabIndex={-1}
    >
      {children}
    </ListItemLabel>
  )

  const Layout = accessory ? ListItemLayoutAccessory : ListItemLayout
  const listItemContent = (
    <Layout
      labelCreator={LabelCreator}
      description={renderedDescription}
      detail={renderedDetail}
      icon={renderedIcon}
      px={itemDimensions.px}
      py={itemDimensions.py}
    >
      {renderedChildren}
    </Layout>
  )

  return (
    <HoverDisclosureContext.Provider value={{ visible: hovered }}>
      <ListItemWrapper
        className={className}
        current={current}
        description={description}
        disabled={disabled}
        focusVisible={focusVisible}
        hovered={hovered}
        keyColor={keyColor}
        ref={itemRef}
        selected={selected}
        {...itemDimensions}
        {...restProps}
      >
        {listItemContent}
      </ListItemWrapper>
    </HoverDisclosureContext.Provider>
  )
}

export const ListItem = styled(ListItemInternal)``
