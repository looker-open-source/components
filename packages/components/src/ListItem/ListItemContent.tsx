/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { CompatibleHTMLProps, DensityRamp } from '@looker/design-tokens'
import React, { FC } from 'react'
import styled, { css, FlattenSimpleInterpolation } from 'styled-components'
import { FocusVisibleProps } from '../utils'
import { ListItemColorProp, ListItemRole, ListItemStatefulProps } from './types'
import {
  listItemBackgroundColor,
  listItemDimensions,
  listItemPadding,
} from './utils'

const Button = styled.button`
  font-family: inherit;
`
const Link = styled.a``
const Div = styled.div``

// Use listItemLabelCSS to target the internal button / link / div CSS of ListItem
export const listItemContentCSS = (style: FlattenSimpleInterpolation) => css`
  > ${Button}, > ${Link}, > ${Div} {
    ${style}
  }
`

/**
 * @deprecated Use `listItemContentCSS` instead
 */
export const listItemLabelCSS = listItemContentCSS

type ListItemContentProps = CompatibleHTMLProps<HTMLElement> &
  ListItemStatefulProps &
  ListItemColorProp &
  FocusVisibleProps & {
    cursorPointer?: boolean
    density?: DensityRamp
    disabled?: boolean
    itemRole?: ListItemRole
  }

const ListItemContentInternal: FC<ListItemContentProps> = ({
  children,
  itemRole = 'button',
  ...props
}) => {
  if (!props.disabled && itemRole === 'link') {
    return <Link {...props}>{children}</Link>
  } else if (itemRole === 'none') {
    return <Div {...props}>{children}</Div>
  }

  return (
    <Button {...props} type="button">
      {children}
    </Button>
  )
}

const itemRoleNoneHeight = (density: DensityRamp = 0) => {
  const { height } = listItemDimensions(density)

  return css`
    min-height: ${height}px;
  `
}

export const ListItemContent = styled(ListItemContentInternal)`
  ${listItemBackgroundColor}

  &:focus {
    ${({ focusVisible, theme }) =>
      focusVisible && `box-shadow: inset 0 0 0 2px ${theme.colors.keyFocus};`}
  }

  align-items: center;
  border: none;
  color: inherit;
  cursor: ${({ cursorPointer }) => (cursorPointer ? 'pointer' : undefined)};
  display: flex;
  flex: 1;
  font-size: inherit;
  font-weight: inherit;
  margin: 0; /* safari has default margin */
  min-width: 0;
  outline: none;
  ${({ density }) => listItemPadding(density)}
  padding-bottom: ${({ itemRole }) => (itemRole === 'none' ? 0 : undefined)};
  padding-top: ${({ itemRole }) => (itemRole === 'none' ? 0 : undefined)};
  ${({ density, itemRole }) =>
    itemRole === 'none' && itemRoleNoneHeight(density)}

  text-align: left;
  text-decoration: none;
  width: 100%;

  &:hover,
  &:focus {
    color: inherit;
    text-decoration: none;
  }
`
