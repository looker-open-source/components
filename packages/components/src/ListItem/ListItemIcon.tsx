/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled, { DefaultTheme, StyledComponent } from 'styled-components'
import {
  color as colorHelper,
  DensityProp,
  SpacingSizes,
} from '@looker/design-tokens'
import { StyledIconBase } from '@styled-icons/styled-icon'
import { IconPlaceholder, IconSize } from '../Icon'
import { listItemDimensions, listItemIconColor } from './utils'

export type ListItemIconProps = DensityProp & {
  color?: string
  disabled?: boolean
  alignStart?: boolean
}

type ListItemIconInternalProps = Omit<ListItemIconProps, 'density'> & {
  iconGap: SpacingSizes
  iconSize: IconSize
}

export const ListItemIcon: StyledComponent<
  'div',
  DefaultTheme,
  ListItemIconProps
> = styled.div.attrs<ListItemIconProps>(
  ({ color, disabled, density = 0, ...props }) => {
    const { iconGap, iconSize } = listItemDimensions(density)

    return {
      ...props,
      color: listItemIconColor(color, disabled),
      iconGap,
      iconSize,
    }
  }
)<ListItemIconInternalProps>`
  align-self: ${({ alignStart }) => (alignStart ? 'flex-start' : 'center')};
  display: flex;
  margin-right: ${({ iconGap, theme }) => theme.space[iconGap]};
  ${colorHelper}

  & > svg,
  ${StyledIconBase}, ${IconPlaceholder} {
    flex-grow: 0;
    flex-shrink: 0;
    height: ${({ iconSize, theme }) => theme.sizes[iconSize]};
    /* justify-content: center; */
    width: ${({ iconSize, theme }) => theme.sizes[iconSize]};
  }
`
