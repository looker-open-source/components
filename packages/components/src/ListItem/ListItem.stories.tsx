/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import type { FC } from 'react'
import React from 'react'
import type { Story } from '@storybook/react/types-6-0'
import { PersonOutline, SubdirectoryArrowLeft } from '@styled-icons/material'
import { DateRange } from '@styled-icons/material-outlined'
import { IconButton } from '../Button'
import { Heading, Text } from '../Text'
import { Grid } from '../Layout/Grid'
import { SpaceVertical } from '../Layout/Space/SpaceVertical'
import { Status } from '../Status'
import { List } from '../List'
import { defaultArgTypes as argTypes } from '../../../../apps/storybook/src/defaultArgTypes'
import { ListItem } from './ListItem'
import type { ListItemProps, ListItemRole } from './types'

const Template: Story<ListItemProps> = args => <ListItem {...args} />

export const Basic = Template.bind({})
const basicArgs = { children: 'List Item' }
Basic.args = {
  ...basicArgs,
}

export const Icon = Template.bind({})
const iconArgs = {
  ...basicArgs,
  icon: <PersonOutline />,
}
Icon.args = {
  ...iconArgs,
}

export const IconColor = () => <ListItem {...iconArgs} color="calculation" />
export const IconStatus = () => (
  <ListItem {...iconArgs} icon={<Status intent="warn" />} />
)
export const IconCustomColor = () => <ListItem {...iconArgs} color="#cc00cc" />
export const IconColorDisabled = () => (
  <ListItem {...iconArgs} color="warn" disabled />
)

const Example: FC<ListItemProps> = ({ children, ...props }) => {
  const args = {
    icon: <PersonOutline />,
    ...props,
  }

  return (
    <SpaceVertical>
      <Heading as="h3">{children}</Heading>
      <List width="100%">
        <ListItem {...args}>Default</ListItem>
        <ListItem {...args} color="key">
          Key
        </ListItem>
        <ListItem {...args} color="calculation">
          Calculation
        </ListItem>
        <ListItem {...args} color="dimension">
          Dimension
        </ListItem>
        <ListItem {...args} color="measure">
          Measure
        </ListItem>
      </List>
    </SpaceVertical>
  )
}

export const ColorComparison = () => (
  <Grid columns={4}>
    <Example>Default</Example>
    <Example selected>Selected</Example>
    <Example selected disabled>
      Selected + Disabled
    </Example>
    <Example hovered>Hover</Example>
  </Grid>
)

export const Detail = Template.bind({})
Detail.args = {
  ...Basic.args,
  detail: 'A Detail',
}

export const Accessory = Template.bind({})
Accessory.args = {
  ...Basic.args,
  detail: {
    content: <IconButton icon={<SubdirectoryArrowLeft />} label="Pivot" />,
    options: { accessory: true },
  },
}

export const HoverDisclosure = Template.bind({})
HoverDisclosure.args = {
  ...Basic.args,
  detail: {
    content: <IconButton icon={<SubdirectoryArrowLeft />} label="Pivot" />,
    options: { hoverDisclosure: true },
  },
}

HoverDisclosure.parameters = {
  storyshots: { disable: true },
}

export const IconAndDetail = Template.bind({})
IconAndDetail.args = {
  ...Basic.args,
  detail: 'A Detail',
  icon: <PersonOutline />,
}

export const Description = Template.bind({})
Description.args = {
  ...Basic.args,
  description: 'A description',
}

export const IconAndDescription = Template.bind({})
IconAndDescription.args = {
  ...Basic.args,
  description: 'A description',
  icon: <PersonOutline />,
}

export const DetailAndDescription = Template.bind({})
DetailAndDescription.args = {
  ...Basic.args,
  description: 'A description',
  detail: 'A detail',
}

export const IconAndDetailAndDescription = Template.bind({})
IconAndDetailAndDescription.args = {
  ...Basic.args,
  description: 'A description',
  detail: 'A detail',
  icon: <PersonOutline />,
}

export const Selected = Template.bind({})
Selected.args = {
  ...Basic.args,
  selected: true,
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const KeyColor = () => <ListItem selected color="key" {...basicArgs} />

export const Link = () => {
  return (
    <ListItem itemRole="link" href="https://google.com" target="_blank">
      ListItem that links to Google
    </ListItem>
  )
}

export const NoRole = () => {
  return (
    <ListItem
      itemRole="none"
      detail={
        <IconButton
          onClick={() => alert('Show me calendar')}
          icon={<PersonOutline />}
          label="Person"
          ml="large"
        />
      }
    >
      ListItem
    </ListItem>
  )
}

const lorem =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'
export const Truncate = Template.bind({})
Truncate.args = {
  ...Basic.args,
  children: lorem,
  description: lorem,
  truncate: true,
}

export const TruncateAndIconAndDetail = Template.bind({})
TruncateAndIconAndDetail.args = {
  ...Truncate.args,
  detail: {
    content: 'Detail',
    options: {},
  },
  icon: <DateRange />,
}

export const TruncateAndIconAndDetailAndAccessory = Template.bind({})
TruncateAndIconAndDetailAndAccessory.args = {
  ...Truncate.args,
  detail: {
    content: <Text fontSize="xsmall">Detail Accessory</Text>,
    options: { accessory: true },
  },
  icon: <DateRange />,
}

const RoleVariant: FC<{ itemRole: ListItemRole; description?: string }> = ({
  itemRole,
  description,
}) => (
  <ListItem itemRole={itemRole} description={description}>
    Hello World
  </ListItem>
)
export const RoleVariants = () => (
  <Grid columns={3}>
    <div>
      <RoleVariant itemRole="button" />
      <RoleVariant itemRole="button" description="Definitely a button" />
    </div>
    <div>
      <RoleVariant itemRole="link" />
      <RoleVariant itemRole="link" description="Definitely a link" />
    </div>
    <div>
      <RoleVariant itemRole="none" />
      <RoleVariant itemRole="none" description="Definitely a none" />
    </div>
  </Grid>
)

export default {
  argTypes,
  component: ListItem,
  title: 'ListItem',
}
