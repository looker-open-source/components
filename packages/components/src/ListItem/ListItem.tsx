/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import React, { forwardRef, Ref, useContext, useRef, useState } from 'react'
import {
  createSafeRel,
  getNextFocusTarget,
  HoverDisclosureContext,
  partitionAriaProps,
  undefinedCoalesce,
  useFocusVisible,
  useWrapEvent,
  useForkedRef,
} from '../utils'
import { IconPlaceholder } from '../Icon'
import { ListItemContext } from './ListItemContext'
import { ListItemContent } from './ListItemContent'
import { ListItemWrapper } from './ListItemWrapper'
import { createListItemPartitions, listItemLabelColor } from './utils'
import { ListItemProps } from './types'

const ListItemInternal = forwardRef(
  (props: ListItemProps, ref: Ref<HTMLLIElement | HTMLDivElement>) => {
    const {
      children,
      className,
      color: propsColor,
      density: propsDensity,
      description,
      detail,
      disabled = false,
      hovered: propsHovered = false,
      href,
      icon,
      itemRole,
      onBlur,
      onClick,
      onKeyDown,
      onKeyUp,
      onMouseEnter,
      onMouseLeave,
      rel,
      role,
      selected,
      tabIndex = -1,
      target,
      truncate,
      ...restProps
    } = props

    const {
      density: contextDensity,
      iconGutter,
      color: contextColor,
    } = useContext(ListItemContext)

    const density = propsDensity || contextDensity

    const color = undefinedCoalesce([propsColor, contextColor])

    const { focusVisible, ...focusVisibleHandlers } = useFocusVisible({
      onBlur,
      onKeyUp,
    })
    const [hovered, setHovered] = useState(propsHovered)

    const handleOnClick = (event: React.MouseEvent<HTMLElement>) => {
      if (itemRole !== 'none' && onClick) {
        onClick(event)
      }
    }
    if (disabled && itemRole === 'link') {
      // eslint-disable-next-line no-console
      console.warn(
        'itemRole="link" and disabled cannot be combined - use itemRole="button" if you need to offer a disabled ListItem'
      )
    }

    if (itemRole === 'none' && onClick) {
      // eslint-disable-next-line no-console
      console.warn(
        'itemRole="none" and onClick cannot be combined - if itemRole="none" is a necessity, assign click behavior directly to ListItem\'s children'
      )
    }

    const wrapperRef = useRef<HTMLLIElement | HTMLDivElement>(null)
    const actualRef = useForkedRef(wrapperRef, ref)

    const [ariaProps, wrapperProps] = partitionAriaProps(restProps)
    const [insideElements, outsideElements] = createListItemPartitions({
      ...props,
      color,
      density,
      icon: icon || (iconGutter ? <IconPlaceholder /> : undefined),
    })

    const statefulProps = {
      color,
      disabled,
      hovered,
      selected,
    }

    const handleWrapperFocus = () => {
      setHovered(true)
    }

    const handleWrapperBlur = (event: React.FocusEvent<HTMLElement>) => {
      const nextFocusTarget = getNextFocusTarget(event)

      if (
        nextFocusTarget &&
        !event.currentTarget.contains(nextFocusTarget as Node)
      ) {
        setHovered(false)
      }
    }

    const handleWrapperMouseEnter = useWrapEvent(
      () => setHovered(true),
      onMouseEnter
    )

    const handleWrapperMouseLeave = useWrapEvent(
      () => setHovered(false),
      onMouseLeave
    )

    return (
      <HoverDisclosureContext.Provider value={{ visible: hovered }}>
        <ListItemWrapper
          className={className}
          color={listItemLabelColor(color, disabled)}
          onBlur={handleWrapperBlur}
          onFocus={handleWrapperFocus}
          onMouseEnter={handleWrapperMouseEnter}
          onMouseLeave={handleWrapperMouseLeave}
          ref={actualRef}
          {...wrapperProps}
        >
          <ListItemContent
            itemRole={itemRole}
            aria-selected={selected}
            cursorPointer={!!(href || onClick)}
            focusVisible={focusVisible}
            href={href}
            onClick={disabled ? undefined : handleOnClick}
            onKeyDown={onKeyDown}
            density={density}
            rel={createSafeRel(rel, target)}
            role={role || 'listitem'}
            target={target}
            tabIndex={tabIndex}
            {...ariaProps}
            {...focusVisibleHandlers}
            {...statefulProps}
          >
            {insideElements}
          </ListItemContent>
          {outsideElements}
        </ListItemWrapper>
      </HoverDisclosureContext.Provider>
    )
  }
)

ListItemInternal.displayName = 'ListItemInternal'

export const ListItem = styled(ListItemInternal)<ListItemProps>``
