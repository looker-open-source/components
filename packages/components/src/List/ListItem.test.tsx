/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import { renderWithTheme } from '@looker/components-test-utils'
import { screen, fireEvent } from '@testing-library/react'

import { ListItem } from './ListItem'

describe('ListItem', () => {
  test('renders children', () => {
    renderWithTheme(<ListItem>who!</ListItem>)
    expect(screen.getByText('who!')).toBeVisible()
  })

  test('renders detail', () => {
    renderWithTheme(<ListItem detail="Is an excellent question">who!</ListItem>)
    expect(screen.getByText('Is an excellent question')).toBeVisible()
  })

  test('renders icon', () => {
    renderWithTheme(<ListItem icon="Beaker">Icon</ListItem>)
    expect(screen.getByText('Icon')).toBeVisible()
  })

  test('renders artwork', () => {
    renderWithTheme(
      <ListItem
        iconArtwork={
          <svg xmlns="http://www.w3.org/2000/svg">
            <title>SVG Title Here</title>
          </svg>
        }
      >
        Artwork
      </ListItem>
    )
    expect(screen.getByTitle('SVG Title Here')).toBeInTheDocument()
  })

  test('is not clickable when disabled', () => {
    const callbackFn = jest.fn()

    renderWithTheme(
      <ListItem itemRole="button" disabled onClick={callbackFn}>
        Item
      </ListItem>
    )

    const item = screen.getByText('Item')
    fireEvent.click(item)

    expect(callbackFn).toHaveBeenCalledTimes(0)
  })

  test('has rel="noopener noreferrer" when it has target="_blank" and no passed-in rel prop value', () => {
    const { getByRole } = renderWithTheme(
      <ListItem itemRole="link" href="https://google.com" target="_blank">
        Link
      </ListItem>
    )

    const item = getByRole('listitem')

    expect(item.nodeName).toBe('A')
    expect(item).toHaveAttribute('target', '_blank')
    expect(item).toHaveAttribute('href', 'https://google.com')
    expect(item).toHaveAttribute('rel', 'noopener noreferrer')
  })

  test('auto appends "noopener noreferrer" to rel prop value when itemRole="link", target="_blank" and rel is not undefined', () => {
    const { getByRole } = renderWithTheme(
      <ListItem
        itemRole="link"
        href="https://google.com"
        rel="nogouda"
        target="_blank"
      >
        Link
      </ListItem>
    )

    const item = getByRole('listitem')

    expect(item.nodeName).toBe('A')
    expect(item).toHaveAttribute('target', '_blank')
    expect(item).toHaveAttribute('href', 'https://google.com')
    expect(item).toHaveAttribute('rel', 'nogouda noopener noreferrer')
  })

  test('does not auto append "noopener noreferrer" to link without target="_blank"', () => {
    const { getByRole } = renderWithTheme(
      <ListItem itemRole="link" rel="nogouda" href="https://google.com">
        Link
      </ListItem>
    )

    const item = getByRole('listitem')

    expect(item.nodeName).toBe('A')
    expect(item).toHaveAttribute('href', 'https://google.com')
    expect(item).toHaveAttribute('rel', 'nogouda')
  })
})
