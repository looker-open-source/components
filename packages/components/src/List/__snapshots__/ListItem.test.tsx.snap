// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem A default ListItem, should be an li 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  margin-bottom: 0.25rem;
}

<li
  className="c0"
>
  ü•ë
</li>
`;

exports[`ListItem handles mouseenter 1`] = `
.c0 {
  font-family: 'Roboto','Noto Sans JP','Noto Sans CJK KR','Noto Sans Arabic UI','Noto Sans Devanagari UI','Noto Sans Hebrew','Noto Sans Thai UI','Helvetica','Arial',sans-serif;
  margin-bottom: 0.25rem;
}

<li
  className="c0"
  onMouseEnter={[Function]}
>
  üê≠
</li>
`;
